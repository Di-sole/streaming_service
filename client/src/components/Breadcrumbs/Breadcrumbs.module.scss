.container {
    display: flex;
    margin-bottom: 24px;

    &__crumbsList {
        display: flex;
        list-style: none;

        li {
            padding-bottom: 3px;

            font-size: 13px;
            line-height: 16px;
            font-weight: 500;

            color: rgba(255, 255, 255, 0.56);

            &::before {
                content: '/';
                margin-left: 4px;
                margin-right: 3px;
                color: #fff;
                border: none;
            }

            &:not(:last-child) {
                cursor: pointer;
            }
        }
    }

    &__crumb {
        list-style: none;
        color: #fff;
        font-size: 13px;
        line-height: 16px;
        font-weight: 500;
        transition-duration: 0.4s;
        border-bottom: 2px solid black;

        &:not(:hover):not(:active) {
            transition-duration: 80ms;
            box-shadow: 0 2px 0 #7e798f, inset 0 0 0 #7e798f;
        }

        &:hover:not(:active) {
            box-shadow: 0 0 0 #7e798f, inset 0 -1.5em 0 #7e798f;
        }
    }
}
