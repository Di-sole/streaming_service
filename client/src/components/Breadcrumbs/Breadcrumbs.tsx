import styles from './Breadcrumbs.module.scss';
import { useTranslation } from 'next-i18next';
import IGenre from '@/models/IGenre';
import { firstCapitalLetter } from '../../utils/functions';
import TextLinkUI from '../UI/links/TextLink/TextLinkUI';
import { useRouter } from 'next/router';
import { useAppSelector } from '../../store/hooks/redux';

interface BreadcrumbsProps {
    path: string[];
    genre?: IGenre;
    tailName?: { name: string; enName: string };
    type?: 'slash' | 'point' | 'pointShort';
}

const Breadcrumbs = ({ path, genre, tailName, type = 'slash' }: BreadcrumbsProps) => {
    const { t } = useTranslation('collection');
    const { locale } = useRouter();
    const { genres, countries } = useAppSelector((state) => state.staticData);

    const localizedPaths = path.map((item) => {
        const splitItem = item.split('-');
        const joinedItem =
            splitItem.length > 1 ? splitItem[0] + splitItem[1][0].toUpperCase() + splitItem[1].slice(1) : splitItem;
        return t(`${joinedItem}`);
    });

    return (
        <div className={styles.container} data-testid={'breadcrumbs'}>
            <ul className={[styles.container__crumbsList, styles[`container__crumbsList_${type}`]].join(' ')}>
                {type !== 'pointShort' && (
                    <TextLinkUI
                        option="bright"
                        href={'/'}
                        className={type === 'slash' ? styles.container__crumb : styles.container__crumb2}
                    >
                        {t('myIvi')}
                    </TextLinkUI>
                )}
                {localizedPaths.map((item, index) => {
                    const last = index === path.length - 1;
                    const ref = path[index]; /*.slice(0, index).join('/')*/
                    /* console.log('localizedPaths', localizedPaths);
                     */

                    if (!last) {
                        if (item !== 'persons') {
                            return (
                                <li key={index}>
                                    <TextLinkUI
                                        option="bright"
                                        href={`/${ref}`}
                                        className={
                                            type === 'slash' ? styles.container__crumb : styles.container__crumb2
                                        }
                                    >
                                        {item}
                                    </TextLinkUI>
                                </li>
                            );
                        }
                    } else {
                        if (genre && tailName) {
                            return (
                                <div key={type}>
                                    <li>
                                        <TextLinkUI
                                            option="bright"
                                            href={`/collections/${genre.enName}`}
                                            className={type === 'slash' ? styles.container__crumb : ''}
                                        >
                                            {firstCapitalLetter(locale === 'ru' ? genre.name : genre.enName)}
                                        </TextLinkUI>
                                    </li>
                                    {type !== 'pointShort' && (
                                        <li className={styles.container__tail}>
                                            {locale === 'ru' ? tailName.name : tailName.enName}
                                        </li>
                                    )}
                                </div>
                            );
                        }
                        if (tailName) {
                            return (
                                type !== 'pointShort' && (
                                    <li key={index}>{locale === 'ru' ? tailName.name : tailName.enName}</li>
                                )
                            );
                        } else {
                            const gener = genres.filter((genre) => genre.enName === item);
                            const country = countries.filter((country) => country.enName === item.split('_').join(' '));
                            if (gener.length > 0)
                                return (
                                    <li key={index}>
                                        {firstCapitalLetter(locale === 'ru' ? gener[0].name : gener[0].enName)}
                                    </li>
                                );
                            if (country.length > 0)
                                return (
                                    <li key={index}>
                                        {firstCapitalLetter(locale === 'ru' ? country[0].name : country[0].enName)}
                                    </li>
                                );
                            return <li key={index}>{item}</li>;
                        }
                    }
                })}
            </ul>
        </div>
    );
};
export default Breadcrumbs;
