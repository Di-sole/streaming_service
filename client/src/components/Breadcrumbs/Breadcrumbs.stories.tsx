import { Meta, StoryObj } from '@storybook/react';
import '../../styles/nullstyle.scss';
import '../../styles/globals.scss';
import './Breadcrumbs.module.scss';
import Breadcrumbs from './Breadcrumbs';

const meta: Meta<typeof Breadcrumbs> = {
    title: 'Links/Breadcrumbs',
    component: Breadcrumbs,
    tags: ['autodocs'],
    argTypes: {
        type: {
            control: false,
        },
    },
    decorators: [
        (Story) => (
            <div style={{ background: '#100e19', padding: '10px' }}>
                <Story />
            </div>
        ),
    ],
};

export default meta;

type Story = StoryObj<typeof Breadcrumbs>;

export const Slash: Story = {
    args: {
        path: ['Фильмы', 'Жанр'],
    },
};

export const Point: Story = {
    args: {
        path: ['Фильмы', 'idMovie'],
        genre: { id: 1, name: 'Жанр', enName: 'Жанр' },
        tailName: { name: 'Название фильма', enName: 'Название фильма' },
        type: 'point',
    },
};

export const PointShort: Story = {
    args: {
        path: ['Фильмы', 'idMovie'],
        genre: { id: 1, name: 'Жанр', enName: 'Жанр' },
        tailName: { name: 'Название фильма', enName: 'Название фильма' },
        type: 'pointShort',
    },
};
