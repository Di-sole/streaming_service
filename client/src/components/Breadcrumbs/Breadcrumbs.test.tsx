import { render, screen } from '@testing-library/react';
import Breadcrumbs from './Breadcrumbs';
import * as reduxHooks from '../../store/hooks/redux';

jest.mock('../../store/hooks/redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const genres = [{ id: 1, name: 'Жанр', enName: 'Genre' }];
const countries = [{ id: 1, name: 'Страна', enName: 'Country' }];

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

jest.mock('next/router', () => ({
    useRouter() {
        return {
            pathname: '',
        };
    },
}));

describe('BREADCRUMBS SNAPSHOTS TESTS', () => {
    test('"type" slash Breadcrumbs should not be changed', () => {
        mockedSelector.mockReturnValue({ genres, countries });
        render(<Breadcrumbs path={['movie', '10']} type="slash" />);

        const div = screen.getByTestId('breadcrumbs');
        expect(div).toMatchSnapshot();
    });

    test('"type" point Breadcrumbs should not be changed', () => {
        mockedSelector.mockReturnValue({ genres, countries });
        render(
            <Breadcrumbs
                path={['movie', '14']}
                genre={{ id: 11, name: 'Жанр', enName: 'Genre' }}
                tailName={{ name: 'Название', enName: 'Name' }}
                type="point"
            />,
        );

        const div = screen.getByTestId('breadcrumbs');
        expect(div).toMatchSnapshot();
    });

    test('"type" pointShort Breadcrumbs should not be changed', () => {
        mockedSelector.mockReturnValue({ genres, countries });
        render(
            <Breadcrumbs
                path={['movie', '13']}
                genre={{ id: 12, name: 'Жанр', enName: 'Genre' }}
                tailName={{ name: 'Название', enName: 'Name' }}
                type="pointShort"
            />,
        );

        const div = screen.getByTestId('breadcrumbs');
        expect(div).toMatchSnapshot();
    });
});
