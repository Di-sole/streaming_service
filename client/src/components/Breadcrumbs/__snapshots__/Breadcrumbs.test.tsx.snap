// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BREADCRUMBS SNAPSHOTS TESTS "type" point Breadcrumbs should not be changed 1`] = `
<div
  class="container"
  data-testid="breadcrumbs"
>
  <ul
    class="container__crumbsList container__crumbsList_point"
  >
    <a
      class="container container_bright container__crumb2"
      href="/"
    >
      myIvi
    </a>
    <li>
      <a
        class="container container_bright container__crumb2"
        href="/movie"
      >
        movie
      </a>
    </li>
    <div>
      <li>
        <a
          class="container container_bright"
          href="/collections/Genre"
        >
          Genre
        </a>
      </li>
      <li
        class="container__tail"
      >
        Name
      </li>
    </div>
  </ul>
</div>
`;

exports[`BREADCRUMBS SNAPSHOTS TESTS "type" pointShort Breadcrumbs should not be changed 1`] = `
<div
  class="container"
  data-testid="breadcrumbs"
>
  <ul
    class="container__crumbsList container__crumbsList_pointShort"
  >
    <li>
      <a
        class="container container_bright container__crumb2"
        href="/movie"
      >
        movie
      </a>
    </li>
    <div>
      <li>
        <a
          class="container container_bright"
          href="/collections/Genre"
        >
          Genre
        </a>
      </li>
    </div>
  </ul>
</div>
`;

exports[`BREADCRUMBS SNAPSHOTS TESTS "type" slash Breadcrumbs should not be changed 1`] = `
<div
  class="container"
  data-testid="breadcrumbs"
>
  <ul
    class="container__crumbsList container__crumbsList_slash"
  >
    <a
      class="container container_bright container__crumb"
      href="/"
    >
      myIvi
    </a>
    <li>
      <a
        class="container container_bright container__crumb"
        href="/movie"
      >
        movie
      </a>
    </li>
    <li>
      10
    </li>
  </ul>
</div>
`;
